### DISCUSSION TITLE
Tip how to make it work on Windows

### DISCUSSION THREAD

**Discussion by jonathanneels** [curif/AgeOfJoy-2022.1#234] [Q&A]
Hi everybody

I've made some changes to the project so it can run on Windows (I recently sold my Meta 2)...
Game starts, movements, location spawning and cabinets with correct textures are working perfectly. 
The directory is also correctly set. 
My error starts when I insert a "coin". 
Screen on the arcade machine turns black and the error
_EntryPointNotFoundException: wrapper_environment_init assembly:<unknown assembly> type:<unknown type> member:(null)
LibretroMameCore.Start (System.String screenName, System.String gameFileName) (at Assets/curif/LibRetroWrapper/LibretroMameCore.cs:376)
is shown._ 

Do note I have the correct roms.
The arcade machine turns black, but all other things keep working fine.

I think the fault must be in file LibretroMameCore.cs. Maybe the reference to  [DllImport ("mame2003_plus_libretro_android")] or is the mame inside the project a different one than https://github.com/libretro/mame2003-plus-libretro/tree/master (Windows compatible?). 

I would love to play and shuffle those cabinets on my pc, like the real deal. 

Help would be tremendous!

Kind regards
Fumetsujo

**Comment by curif** [comment]
Please, you need to do it in the 0.4 branch. Master is outdated; I push to master only when I publish in Itchio. Please create a new branch starting from 0.4

I've been thinking about publishing a Windows version, but the challenge is the player control. The hands aren't designed to be used with a keyboard or a control. In the new version, you can use the WASD combo to move the player.
Another challenge could be the libretro library too, but I believe it should be easy to put it to work.

Regards
Curif.

**Reply by jonathanneels** [reply]
Thank you for your feedback! 

I think I know where the fault lies. 
In LibretroMameCore.cs the ddlimport refers to __internal. I have no idea what that __internal must be called on the Windows platform.
I'll try in the 0.4 build. Suggestions are more than welcome!

Kind regards
Fumetsujo

**Reply by curif** [reply]
In short, `__internal` means that Unity looks for compiled C code. I moved a lot of functions from c# to C code in order to speed up some critical CPU-consuming functions, like image conversion and audio.

You can find the C code in the LibretroWrapper/cwrapper folder.

**Reply by jonathanneels** [reply]
Ok, thank you for the feedback. I'll keep you posted when I made some progress :).

**Comment by Loafmeister** [comment]
I was shocked to see this @jonathanneels , interesting work. Is there any progress?
